### ISSUE TITLE
Using Epoll instead of NIO

### ISSUE THREAD

**Issue by sschepens** [AsyncHttpClient/async-http-client#894]
Is there a reason why it's not allowed to specify an Epoll WorkerGroup and also why can't SocketChannel classes be configured?
Seems like if one could set a EpollWorkerGroup and EpollSocketChannel everything else should work the same, but using epoll instead of nio.

**Comment by hepin1989** [comment]
maybe using something like 
```
    val native = if (Platform.isLinux && preferNative) true else false
```

**Comment by slandelle** [comment]
Contributions welcome :)

@hepin1989 No reason to limit this to Linux. It should work on all platforms where there's a tcnative implementation available.

**Comment by hepin1989** [comment]
@slandelle but the current epoll transport just works on linux.
for the tcnative ,should be what as you said. I just Ctrl+C ,Ctrl+V,:)

**Comment by slandelle** [comment]
@hepin1989 What about the jars that are available for Windows and OSX? http://search.maven.org/#search%7Cga%7C1%7Ctcnative

> I just Ctrl+C ,Ctrl+V,:)
From where?

**Comment by hepin1989** [comment]
@slandelle haha,seems like we are talking about different things,from my own code,I am implementing akka io extension via netty4.

what I said was this https://github.com/netty/netty/tree/master/transport-native-epoll
:)
haha ,thanks I will take a look at the tcnative,I have not looked into that,thanks.

**Comment by slandelle** [comment]
Oh right! I had missed that, I thought that tcnative was also providing native epoll.

Then, if we want to start supporting native implementations, we probably want to also support the OpenSSL based SSLEngine.

@sschepens @hepin1989 Feeling like contributing?

**Comment by sschepens** [comment]
@slandelle I'll see if i cant make this, It seems to me that what we should do first is remove [this](https://github.com/AsyncHttpClient/async-http-client/blob/master/providers/netty4/src/main/java/org/asynchttpclient/providers/netty/channel/ChannelManager.java#L170-L171).
Then make [these](https://github.com/AsyncHttpClient/async-http-client/blob/master/providers/netty4/src/main/java/org/asynchttpclient/providers/netty/channel/ChannelManager.java#L173-L176) uses of NioSocketChannel.class use configuration, like:
```java
String socketChannelClass = nettyConfig.getSocketChannelClass();
plainBootstrap = new Bootstrap().channel(socketChannelClass).group(eventLoopGroup);
secureBootstrap = new Bootstrap().channel(socketChannelClass).group(eventLoopGroup);
webSocketBootstrap = new Bootstrap().channel(socketChannelClass).group(eventLoopGroup);
secureWebSocketBootstrap = new Bootstrap().channel(socketChannelClass).group(eventLoopGroup);
``` 
Seems ok?

**Comment by hepin1989** [comment]
@sschepens Ok you have take it.
I am using typesafe config and with an config section with ```prefer-native = ok```.

the channel could be replaced via ChannelFactory here I think,and we need to test the Platform too.

**Comment by slandelle** [comment]
@sschepens

>  I'll see if i cant make this, It seems to me that what we should do first is remove this restriction.

Or maybe turn it into checking the user is not trying to use Oio.

@hepin1989 

AHC2 currently ships config 1.2.
I think that's a mistake, because of the recent breaking change of config 1.3 targeting Java 8. There's a chance we quickly hit compatibility issues. I'm going to drop it.

The `ChannelManager` part is the easy stuff.
The more complex stuff is to decide how the behavior should be triggered and how the extra dependencies should be added to the project.

IMHO, extra deps shouldn't be required (Windows and OSX users don't want Linux specific libs), it should be up to the user to add them. And if they're not here, code shouldn't crash at runtime with CNFE, ie it has to be hidden behind a strategy.

**Comment by hepin1989** [comment]
yeah ,then that should be an stub,if have the native transport,then use it for linux,otherwise fallback to NIO right?thanks for the detail .

**Comment by sschepens** [comment]
@slandelle is there a new release planned soon? I can see 1.9.x only supports netty 3 and it's configured as a dependency right? This is a bummer!
Master seems to not require this dependency and leave it up to user to define it.
By the way, are you aware that Netty 4.0.28 just got a ChannelPool implementation? and a ChannelPoolMap as well, maybe we could ditch current implementations to favor those!

**Comment by slandelle** [comment]
> is there a new release planned soon?

I can release an alpha version right now so people can start testing, but I don't think master is stable yet, all the more the Netty 4 provider.

> Master seems to not require this dependency and leave it up to user to define it.

Master is modularized. You have to pull either the Netty 3 or the Netty 4 provider.

> By the way, are you aware that Netty 4.0.28 just got a ChannelPool implementation? and a ChannelPoolMap as well

Yes, of course :)

> maybe we could ditch current implementations to favor those!

`SimpleChannelPool` has a different logic, as it opens channels. And it lacks some features, like the ability to specify the local address for IP spoofing. So, it's maybe feasible, but replacing would need some work.

**Comment by sschepens** [comment]
> SimpleChannelPool has a different logic, as it opens channels. And it lacks some features, like the ability to specify the local address for IP spoofing. So, it's maybe feasible, but replacing would need some work.

But can't you do this sort of things via the `bootstrap.localAddress()`?
What we should be using is a class extending from `AbstractChannelPoolMap` which allows to define how to instantiate a new `ChannelPool` per key, and a `SimpleChannelPool` takes a bootstrap, so we could and should define a new `Bootstrap` for each key in the `AbstractChannelPoolMap` of course using the same `EventLoopGroup`, and with `Bootstrap` we could the define remote and local address we need.

**Comment by slandelle** [comment]
Do you want to give it a try, then?

Regarding `AbstractChannelPoolMap`, I was wondering why @normanmaurer didn't go with Java 8 CHM API (either regular CHM when running Java 8, or the forked CMHV8) so he could use `computeIfAbsent` instead of destroying unnecessary pools.

**Comment by sschepens** [comment]
@slandelle I left two pull requests, could you give them a look?
Regarding connection pools I'll see if I have time to implement it.

**Comment by slandelle** [comment]
@sschepens Merged, thanks!

**Comment by sschepens** [comment]
@slandelle I've been checking `ChannelManager`, `ChannelPool` and etc, what I see as a complication is that all logic of creating channels and connections is not actually in them, or so it seems. Logic is very scattered and makes it difficult to trace where things are used.
What's weird is that `ChannelPool` is not the one in charge of initializing channel bootstraps and so on, Nett's `ChannelPool` is thought to have the logic to handle this.

I'm trying to give this a go, but it's gonna take time for me to get used to how the code works and to try to move the code around and not break anything.
So patience...

**Comment by slandelle** [comment]
> What's weird is that ChannelPool is not the one in charge of initializing channel bootstraps and so on, Netty's ChannelPool is thought to have the logic to handle this.

I agree, that was also my reaction when I first saw it.

But then, you have to remember that AHC was the first asynchronous HTTP client on the JVM, so yes, there's many things that would need to be redesigned. I spent quite a lot of time cleaning up the original code so it could make sense to me and it could then be redesigned. Hopefully, with a little help... ;-)

**Comment by slandelle** [comment]
Seems only partly implemented in 16997aae96a58d4f78a5597907c3a9ab55228c51.
It looks like `NioEventLoopGroup` should also be replaceable by `EpollEventLoopGroup`.

**Comment by slandelle** [comment]
I finally went with the `preferNative` option:
* when EventLoopGroup is set, automatically deduce the Socket class from it
* otherwise, use `preferNative`

Classes are loaded by reflection.
